### ISSUE TITLE
Not sure if this is a vagrant-libvirt issue but SD card is not correctly read. blk_update_request: I/O error.

### ISSUE THREAD

**Issue by Folaht** [vagrant-libvirt/vagrant-libvirt#1346]
### Steps to reproduce
1. `vagrant init safenetwork-community/manjaro-arm-installer`
2. `vagrant up`
3. Open virt-manager console
4. Add SD card reader
5. Go back to virt-manager console

### Expected behaviour

SD card is recognized without issue.

### Actual behaviour

![image](https://user-images.githubusercontent.com/9342090/132958288-453cd7ca-c573-4885-baec-ea580e15ca58.png)

### System configuration
**OS/Distro version:**:

5.4.143-1-MANJARO x86_64

**Libvirt version:**

1:7.6.0-1

**Output of `vagrant version; vagrant plugin list`:**

2.2.18

**Attach Output of `VAGRANT_LOG=debug vagrant ... --provider=libvirt >vagrant.log 2>&1`**

https://gist.github.com/Folaht/c86f2cb1a346d3d1950d3565fb5d2f15

**Comment by abbbi** [comment]
After "vagrant-up" the vagrant-libvirt plugin is not responsible for any changes done to the virtual machine via virt-manager. You will also lose them after destroying the virtual machine. You have to check why libvirt doesnt pass through the device. If it is an sd card reader it might aswell be a issue of the box you are spinning up.

**Comment by electrofelix** [comment]
Assuming based on the subsequent thumbs up and nothing further that this is resolved now? Marking for closing automatically unless there is anything more that we should be looking at here.

**Comment by Folaht** [comment]
It's not resolved at all. I tried the virtualbox version of this box and it works fine.

**Comment by electrofelix** [comment]
@Folaht Reading the box description, I wonder if this is a known issue?

https://app.vagrantup.com/safenetwork-community/boxes/manjaro-arm-installer:
![image](https://user-images.githubusercontent.com/912324/134880270-b1c50aed-4744-4d41-b125-ef008c713525.png)

**Comment by Folaht** [comment]
The issue is known to me. I created the box.  
The issue also present in Zelec/manjarolinux.

**Comment by electrofelix** [comment]
Start of the process seems to be working for me, took a while for me to find my sdcard reader.

I brought the VM up, did a pass through of the USB sdcard reader (ensuring I had unmounted the sdcard from the host), restarted the VM using `vagrant halt && vagrant up`.

![image](https://user-images.githubusercontent.com/912324/134886091-593dc046-317e-4f23-9463-871562124319.png)

I did pre-format the device as a single partition ext 4 via my host system before hand, but it appears to reformat it subsequently suggesting it's able to access the sdcard for me?

![image](https://user-images.githubusercontent.com/912324/134885702-951334df-8918-4093-9791-c6d4a70112bc.png)

Waiting to see if it completes at the moment.

**Comment by electrofelix** [comment]
So far no buffer issues reported:
![image](https://user-images.githubusercontent.com/912324/134886429-1bde4749-4a65-482a-a8c7-bbf18603d058.png)

**Comment by electrofelix** [comment]
@Folaht looks to me like this works with libvirt as is?

![image](https://user-images.githubusercontent.com/912324/134887749-16b4b0be-4a90-4d2a-a0a7-cde3e0267856.png)

**Comment by Folaht** [comment]
Looks like to me as well.  
You just showed me something I didn't do.  
I'll do one test and then close.

**Comment by electrofelix** [comment]
@Folaht what was it I did differently? The restart of the VM?

Just done some experimenting, there is definitely something not quite right with how the usb host controller is set up by vagant-libvirt, which is probably the source of the issues with the pass through.

**Comment by Folaht** [comment]
Yeah, the usb is set to pcii-hxtrt or something?

I'm right now testing for the third time to see if Manjaro OS will be installed on the USB using the box.
I did some idiot things the first two times (choosing vda, choosing btrfs).

**Comment by Folaht** [comment]
Great it works!  
Shall we close this issue and start a new one?
I'll try to get the actual USB model it is set to.

**Comment by electrofelix** [comment]
@Folaht by default vagrant-libvirt is using piix3-uhci, when adding with virt-manager it uses ich9-ehci1, but it'll also add another 3 controller devices using models ich9-uchi1, ich9-uchi2, & ich9-uchi3.

Trying to match this runs into issues around controllers needing to specify the master device which is not supported by vagrant-libvirt, however in my case, switching the model to echi seems to be sufficient.

The following appears to work for me in that the SD card device becomes fully accessible inside the VM.
```
Vagrant.configure("2") do |config|
  config.vm.box = "safenetwork-community/manjaro-arm-installer"
  config.vm.provider :libvirt do |libvirt|
    libvirt.usb_controller :model => "ehci"
    libvirt.usb :vendor => "0x1908", :product => "0x0225"
  end
end
```

You can try replacing the vendor / product IDs with those corresponding to your sdcard reader and hopefully it should pass through enough that you can check via `sudo fdisk -l /dev/sda` to confirm that it's fully visible to the guest.

The full list of models are:
* piix3-uhci
* piix4-uhci
* ehci
* ich9-ehci1
* ich9-uhci1
* ich9-uhci2
* ich9-uhci3
* vt82c686b-uhci
* pci-ohci
* nec-xhci
* qusb1
* qemu-xhci

Apparently ochi & uhci are USB 1 only, ehci is USB 2 only, and xhci is supposed to support all with a single stack. So if we can confirm that the default of nec-xhci or qemu-xhci work out of the box for both of us using this, then might switch the default to one of these models instead of the current one.

**Comment by electrofelix** [comment]
Either of nec-xhci or qemu-xhci appear to work as needed for me, and there is no need to umount the device before starting up the VM, it just works.

**Comment by Folaht** [comment]
qemu-xhci works for me.

**Comment by electrofelix** [comment]
Maybe hold off for a PR to go up to change the default model to being qemu-xhci

**Comment by Folaht** [comment]
> make it work with just libvirt.usb :vendor => "<id>", :product => "<id>"

How is that simpler than doing it via GUI?

**Comment by electrofelix** [comment]
When I used the GUI I happened to change the USB model and also restart the VM. Maybe I missed that it's possible to do this by adding the USB controller first with the right model and then add the USB device. However that's still 2 steps and a specific order.

In this case given it would be possible to construct the correct `libvirt.usb` line and have it be inserted automatically with some ruby code based on being provided a suitable device identifier.

Whether it's easier to use the GUI or Vagrantfile can be argued, however I feel that by default if the Vagrantfile is used it shouldn't take some investigation to discover that you must also explicitly set the usb controller model to `qemu-xhci`. Instead I can add some code that if the usb controller model is not defined, and any usb device is specified for pass through or redirection, then to insert a sensible default.
